langchain/evaluation/run_evaluators/message_run_evaluator.py

```diff
@@ -0,0 +1,138 @@
+
+
+class MessageRunMapper(Serializable):
+    """Extract items to evaluate from run object"""
+
+    @property
+    def output_keys(self) -> List[str]:
+        """The keys to extract from the run."""
+        return ["prediction", "input"]
+
+    @abstractmethod
+    def map(self, run: Run) -> RunMapping:
+        """Maps the Run to a dictionary."""
+
+    def __call__(self, run: Run) -> RunMapping:
+        """Maps the Run to a dictionary."""
+        if not run.outputs:
+            raise ValueError(f"Run {run.id} has no outputs to evaluate.")
+        return self.map(run)
+
+
+class MessageExampleMapper(Serializable):
+    """Map an example, or row in the dataset, to the inputs of an evaluation."""
+
+    reference_key: Optional[str] = None
+
+    @property
+    def output_keys(self) -> List[str]:
+        """The keys to extract from the run."""
+        return ["reference"]
+
+    def map(self, example: Example) -> ExampleMapping:
+        """Maps the Example, or dataset row to a dictionary."""
+        if not example.outputs:
+            raise ValueError(
+                f"Example {example.id} has no outputs to use as a reference."
+            )
+        if self.reference_key is None:
+            if len(example.outputs) > 1:
+                raise ValueError(
+                    f"Example {example.id} has multiple outputs, so you must"
+                    " specify a reference_key."
+                )
+            else:
+                output = list(example.outputs.values())[0]
+                return {
+                    "reference": output if isinstance(output, BaseMessage) else messages_from_dict([output])[0]
+                }
+        elif self.reference_key not in example.outputs:
+            raise ValueError(
+                f"Example {example.id} does not have reference key"
+                f" {self.reference_key}."
+            )
+        output = example.outputs[self.reference_key]
+        return {"reference": output if isinstance(output, BaseMessage) else messages_from_dict([output])[0]}
+
+    def __call__(self, example: Example) -> ExampleMapping:
+        """Maps the Run and Example to a dictionary."""
+        if not example.outputs:
+            raise ValueError(
+                f"Example {example.id} has no outputs to use as areference label."
+            )
+        return self.map(example)
+
+
+class ChatModelMessageRunMapper(MessageRunMapper):
+    """Extract items to evaluate from run object."""
+
+    @staticmethod
+    def extract_inputs(inputs: Dict) -> List[BaseMessage]:
+        if not inputs.get("messages"):
+            raise ValueError("Run must have messages as inputs.")
+        if "messages" in inputs:
+            if isinstance(inputs["messages"], list) and inputs["messages"]:
+                if isinstance(inputs["messages"][0], BaseMessage):
+                    return messages_from_dict(inputs["messages"])
+                elif isinstance(inputs["messages"][0], list):
```

Comment:
even if we don't merge this is PR in, we should cherry pick this to the LLMStringRunMapper cc @hinthornw @nfcampos

Reply:
Cherry picked